src/compiler/emitter.ts

```diff
@@ -2061,14 +2051,28 @@ namespace ts {
 
         function emitSourceFile(node: SourceFile) {
             writeLine();
-            emitBodyIndirect(node, node.statements, emitSourceFileWorker);
+            // If the node is a sourceFile and it has prologueDirective (statmentOffSet is not zero) that is synthesize
+            // We will need to emit detached comment here because emitPrologueDirective will not emit comments of such prologue directive
+            const statements = node.statements;
+            // Emit detached comment if there iare no prologue directives or this is a synthesized prologue directives.
+            // The synthesized node will have no leading comment so some comments may be missed.
+            const shouldEmitDetachedComment = statements.length === 0 ||
+                !isPrologueDirective(statements[0]) ||
+                (isPrologueDirective(statements[0]) && nodeIsSynthesized(statements[0]));

```

Comment:
No need to duplicate the `isPrologueDirective` test, as the alternate condition will only be evaluated if the statement is a prologue directive due to shortcutting.

Also, Why not nest the more exhaustive checks behind the check for the presence of `emitBodyWithDetachedComments` so that anyone consuming the `createPrinter` API doesn't have to pay the cost?